server/services/ai.js

```diff
@@ -28,6 +53,179 @@ module.exports = class AIService {
         player.credits = Math.max(0, player.credits);
     }
 
+    async _setupAi(game, player) {
+        player.researchingNext = 'random'; // Set up the AI for random research.
+        
+        // Make sure all stars are marked as not ignored - This is so the AI can bulk upgrade them.
+        const playerStars = this.starService.listStarsOwnedByPlayer(game.galaxy.stars, player._id);
+
+        for (let star of playerStars) {
+            this.starService.resetIgnoreBulkUpgradeStatuses(star);
+        }
+
+        // Clear out any carriers that have looped waypoints.
+        this.carrierService.clearPlayerCarrierWaypointsLooped(game, player);
+
+        const starGraph = this._computeStarGraph(game, player, playerStars);
+
+        // Star systems with computed score based on distance to enemy
+        const starScores = this._computeStarScores(game, player, playerStars, starGraph);
+
+        // Graph of carrier movements for logistics
+        const logisticsGraph = this._createLogisticsGraph(starGraph, starScores);
+        
+        const carrierLoops = this._computeCarrierLoopsFromGraph(logisticsGraph, playerStars);
+
+        const logisticsOrders = this._createCarrierOrders(carrierLoops);
+
+        player.scheduledOrders = logisticsOrders;
+    }
+
+    _computeStarGraph(game, player, playerStars) {
+        const hyperspaceRange = this.distanceService.getHyperspaceDistance(game, player.research.hyperspace.level);
+
+        const starGraph = new Map();
+
+        playerStars.forEach((star, starIdx) => {
+            const reachableStars = new Set();
+
+            playerStars.forEach((otherStar, otherStarIdx) => {
+                if (this.distanceService.getDistanceBetweenLocations(star.location, otherStar.location) <= hyperspaceRange) {
+                    reachableStars.add(otherStarIdx);
```

Comment:
If `otherStarIdx` is already in the set then you already know that the star is within hyperspace range so no need to check the distance again.

A good example of this kind of thing is the `starService.filterStarsByScanningRange` function which uses `scanning` instead of hyperspace range. Same principle I think.

Reply:
Sorry, I don't think I get it. The graph needs to be reflective, though-- And a quick number comparison is probably faster thatn fetching the corresponding set of  `otherStar` and checking inside it, or what else were you talking about?